### ISSUE TITLE
Please add notion.so support :)

### ISSUE THREAD

**Issue by xhr15** [bitcrowd/tickety-tick#163]
It would be lovely to use tickety-tick with notion.so

**Comment by pmeinhardt** [comment]
We could maybe do a lunch coding session maybe to spread a bit of knowledge on how to work on Tickety-Tick, some general JS tooling etc. 🙂

**Comment by xhr15** [comment]
it turned out notion is:
1) hard to parse
2)saying that an API is on the roadmap, but there is no progress

Therefore we are dropping notion support for now

**Comment by pmeinhardt** [comment]
Notion is likely back for a new project, so let's see whether we can come up with a way of supporting it. ✌️

**Comment by pmeinhardt** [comment]
So, here's my findings so far:

- Looks like we cannot really tell what a "ticket" is in notion, because they keep everything super generic and don't have a notion (pun intended) of what is a "ticket".
- However, it looks like everything is basically a "page", so we could generate commit messages and branch names based on pages.

Which information do we need?

- ticket identifier
- ticket title
- (optional) type, like "feature", "bug"…
- (optional) description
- (optional) URL

How could we go about it?

1. Starting from the URL, we can derive the page ID (UUID)

    - When directly on the ticket page, the URL looks like

      `https://www.notion.so/<organization name>/<slugified title>-<page uuid, no dashes>`

      Turns out you can drop the `slugified title` and the URL still works. This would mean we can construct a shorter (optional) URL to include in the ticket information.

    - When opened in an overlay, the URL follows this pattern

      `https://www.notion.so/<organization name>/<parent uuid, no dashes>?v=<version?>&p=<page uuid, no dashes>`

    - So the "proper" page UUID can be constructed by inserting the dashes in the appropriate places, e.g.

      `96ec637de4b04a5eacf38d4d9a1b2e4b` becomes `96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b`

2. We then send a `POST` request to `https://www.notion.so/api/v3/getRecordValues` (or, since we're on the same domain, `/api/v3/getRecordValues`)

    - The request body should contain:

      ```json
      {
        "requests": [
          { "table": "block", "id": "96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b" }
        ]
      }
      ```

    - This will return a JSON response like the following:


      ```json
      {
        "results": [
          {
            "role": "editor",
            "value": {
              "id": "96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b",
              "version": 111,
              "type": "page",
              "properties": {
                "title": [
                  [
                    "Create /api/v1/pages endpoint"
                  ]
                ],
                // …
              },
              "content": [
                "9744d655-a632-4d24-bd47-4348aaab99d6",
                "80c4c1f0-3131-4c50-8d43-9d20cff21824",
                "47842d57-eb5f-40d9-b35f-04e388ffd66c",
                // …
                "0980397d-9f93-4350-9ea2-e08bffb6478b",
                "d3933796-6b33-4771-bf66-36b7caa2f9b7"
              ],
              // …
            }
          }
        ]
      }
      ```

    - Side note: The request might need include cookies for authentication, which shouldn't be a problem as the extension is executing in the context of the page

3. So far we have the ticket identifier (page UUID). We might want to encode this to make it feasible to include in commit messages (or actually change the commit message format to not include ticket identifiers in the subject line).

   From the response, we also get the raw ticket title (`results..value.properties.title[0][0]`).

   With the UUIDs from the `"content"` we could derive more information. But I'd just not do that for now.

Looks like this could be indeed feasible to implement. :v:

**Comment by klappradla** [comment]
Thanks a 🚚 for the investigation @pmeinhardt 🙏 🙏 🙏 

Hacked something together 👇 
<img width="1370" alt="Screenshot 2020-04-08 at 19 06 02" src="https://user-images.githubusercontent.com/3491815/78812748-318f8500-79cc-11ea-8144-8b40417b4dcd.png">

With that I would generate:
**branch:**

```
task/96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b-create-api-v1-pages-endpoint
```

**commit message:**
```
[#96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b] Create…

…/api/v1/pages endpoint
```

😹 the UUID is a bit bulky as a ticket identifier... But I up until now could not find anything better in the response. Did anything else cross your 👀 while investigating or should we just live with the bulky thing for now?

**Comment by pmeinhardt** [comment]
You are awesome @klappradla! 🤘❤️ 

> 😹 the UUID is a bit bulky as a ticket identifier... But I up until now could not find anything better in the response. Did anything else cross your 👀 while investigating or should we just live with the bulky thing for now?

My initial thought while quickly throwing together the concept was that a UUID is just 128 bits of data which – if encoded with an encoding that is able to put more bits into fewer characters (say base58) – could be shortened by a fair amount. 💭 

However, we'd lose the ability to just quickly use the ID from the commit to find the corresponding ticket (you'd have to decode it first). ⚡️ So, in short, I think we should **keep the UUIDs**.

However, we might want to **consider not putting the ticket ID in the subject line of commits**. I started to wonder about this anyway and thought we might want to update our default formats. 🤔 

The current default commit message template is:

```
[#{id}] {title}

{description}

{url}
```

Maybe something like the following would actually make more sense?

```
{title}

{description}

Closes #{id}.

{url}
```

This would also mean GitHub issues would automatically be closed, which seems like a nice default?

**Comment by pmeinhardt** [comment]
I actually just had another thought and it actually seems to work. 💡 

I was thinking about how Git allows "Short SHA-1" using only the first few characters of the hash (at least 4, usually 7), as long as they're unambiguous.

https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection

This actually seems to work for Notion.so and their UUIDs as well.

For instance, in your example above, the UUID and full URL are:

```
96ec637d-e4b0-4a5e-acf3-8d4d9a1b2e4b
https://www.notion.so/bitcrowd/Create-api-v1-pages-endpoint-96ec637de4b04a5eacf38d4d9a1b2e4b
```

However, you'll get to the same page using just this (the first "segment" of the UUID in standard format until the first dash):

```
96ec637d
https://notion.so/bitcrowd/96ec637d
```

**Comment by pmeinhardt** [comment]
Argh. Nope. Scratch that. They just redirect you to the last page you've visited. 🤦‍♂
